add local s3 storage emulation with minio

gumaerc opened the pull request:
#### Pre-Flight checklist

- [x] Testing
  - [x] Code is tested
  - [ ] Changes have been manually tested
- [x] Settings
  - [x] New settings are documented and present in `app.json`
  - [x] New settings have reasonable development defaults, if applicable
  - [x] Opened issue for DevOps regarding necessary configuration changes to deployed environments

#### What are the relevant tickets?
Closes https://github.com/mitodl/ocw-studio/issues/1420

#### What's this PR do?
This rather large PR adds new docker containers to the local docker-compose configuration to run [Minio](https://github.com/minio/minio) locally. Minio emulates Amazon S3, allowing you to isolate the `ocw-studio` publishing process locally for testing without having to touch any S3 buckets actually in AWS. Any initialization of the `boto3` library in Python code will now point itself at Minio if `settings.ENVIRONMENT` is set to "dev." Management commands that upsert pipelines to Concourse also now use a new utility function to strip out lines based on if your environment is "dev" or not. Dev pipelines will override the AWS endpoint URL to point to Minio and skip things like the Slack webhooks, Open Discussions webhooks, automatic triggering, etc.

#### How should this be manually tested?
 - Read the updated readme and set up the Minio env variables based on it. Also look at the updated Concourse section and the readme overall and make sure the changes make sense. Also, make sure you have Github integration set up with a custom organization.
 - Spin up `ocw-studio` with `docker-compose up`
 - Pull and restore a backup of the production database:
```
pg_dump PROD_DB_URL > ./dumps/prod_TODAY_DATE.sql
docker-compose rm -sv db
docker-compose up -d db
docker-compose exec -T -e PGPASSWORD=postgres db psql -h db -U postgres < ./dumps/prod_TODAY_DATE.sql
```
(alternatively, make sure you have an `ocw-www` site as well as an `ocw-course` site)
 - In the logs, you will see the `create-buckets` container spin up, create some containers and set permissions and then exit. This is normal and runs to create buckets in Minio.
 - Upsert the theme assets pipeline to your local Concourse with `./manage.py upsert_theme_assets_pipeline, log into Concourse at http://concourse:8080, start a theme assets build and wait for it to succeed.
 - Browse to your `ocw-www` site at http://localhost:8043/sites/ocw-www. Publish the site to draft.
 - Browse to http://localhost:8080 and log into Concourse. Find the `ocw-www` live pipeline and make sure it completes successfully.
 - Once the pipeline is complete, visit http://localhost:8044 and verify you see the OCW home page. Repeat the last 3 steps for live and verify the same thing but at http://localhost:8045
 - Create a site using the `ocw-course` starter. Visit the Resources section, click the Google Drive link and upload an example file. Click the "Sync w/ Google Drive" button and wait until it succeeds. Visit http://localhost:9001 and login using the credentials detailed in the readme. Browse to the storage bucket and verify that your file was successfully uploaded.
 - Go back to your `ocw-course` site and publish it to draft.
 - Go back to Concourse and locate the draft pipeline for your course site, make sure it completes.
 - Once the pipeline is complete, visit http://localhost:8044/courses/your_course and verify you see the OCW home page. Repeat the last 3 steps for live and verify the same thing but at http://localhost:8045/courses/your_course
 - Smoke test any other functionality in `ocw-studio` that you can

#### Any background context you want to provide?
This is obviously a pretty big PR that touches a lot of existing functionality, so please reach out with any questions while reviewing.

mbertrand reviewed (commented):
Some initial feedback, I've gotten individual courses to publish successfully but not the home page (ocw-www)

mbertrand commented on README.md:
````diff
@@ -225,18 +195,49 @@ If you wish to use a github app,
+AWS_ARTIFACTS_BUCKET_NAME=ol-eng-artifacts
+RESOURCE_BASE_URL_DRAFT=https://draft.ocw.mit.edu
+RESOURCE_BASE_URL_LIVE=https://ocw.mit.edu
+```
````
These need to be set to functioning URLs for the ocw-www pipeline to complete successfully.

```
OCW_STUDIO_LIVE_URL=https://ocw.mit.edu
OCW_STUDIO_DRAFT_URL=https://draft.ocw.mit.edu
```

or maybe the pipeline variable replacement code could be adjusted to set the value of `static-api-base-url` based on `RESOURCE_BASE_URL_DRAFT` and `RESOURCE_BASE_URL_LIVE` instead when `ENVIRONMENT=dev`?

gumaerc commented on content_sync/pipelines/concourse.py:
```diff
@@ -364,17 +382,17 @@ def upsert_pipeline(self):  # pylint:disable=too-many-locals
                 private_key_var = ""
             build_drafts = "--buildDrafts" if pipeline_name == VERSION_DRAFT else ""
 
+            pipeline_template = f"definitions/concourse/site-pipeline{dev_suffix}.yml"
```
This is a great idea!

gumaerc commented:
@mbertrand This is ready for another look now.  Here is what I did:

1. Implemented some new utils in `content_sync/utils.py` to achieve the comment delimiting system we need for stripping out dev / non dev config lines from pipeline templates along with some tests.  I was very careful to make sure that these are thoroughly tested, but you should go through these with a fine-toothed comb.  We certainly don't want lines being unintentionally stripped, so this functionality needs to be bulletproof.  All of the upsert code and tests in `concourse.py` was updated to process the pipeline in this way and a lot of refactoring was done.
2. Added some tweaks to the `nginx.conf` for dev so that you are automatically redirected to `/index.html` when hitting the root of a folder.
3. Added the `STATIC_API_BASE_URL` setting as `dev_only` which will set `STATIC_API_BASE_URL` to this value on all pipelines you upsert if `is_dev` is true, otherwise falling back to `OCW_STUDIO_DRAFT_URL` or `OCW_STUDIO_LIVE_URL` if it's not set based on what you're publishing.  This will allow you to set `STATIC_API_BASE_URL` to https://ocw.mit.edu to source all your static JSON content from the live site while putting http://localhost:8045 for the other settings so that your publish drawer URLs point directly to your Minio hosted content.
4. The readme was updated detailing everything you touched on as well as some other points.

gumaerc commented:
@mbertrand Side note, while developing tests for this I noticed some potential issues with the way we mock our Django settings values, at least with the `content_sync` tests.  The test fixture method we're using works fine if you're simply asserting the effects of the functions the test case runs, but there seems to be an issue when you introduce mocker.  When I was experimenting with adding extra instance vars, the mocker patch call would seemingly return not only the wrong values, but the values from my `.env` file.  I'll need to do more experimentation to see exactly what the problem is here.

mbertrand reviewed (changes requested):
Functionality works great

gumaerc commented on config/nginx.conf:
```diff
@@ -25,3 +25,69 @@ server {
+        # Pass to the Minio publish bucket
+        proxy_pass http://s3:9000/${AWS_PUBLISH_BUCKET_NAME}/;
+    }
+}
```
@mbertrand The nginx.conf file is only used in dev I believe.  The comment at the top says "This is the version used in development environments."  It's pulled in here in `docker-compose.yml`:

```
  nginx:
    volumes:
      - ./config/nginx.conf:/etc/nginx/templates/web.conf.template
      - ./:/src
```

So, I believe that Heroku doesn't use this file at all, but if you know otherwise definitely let me know.

gumaerc commented on content_sync/utils.py:
```diff
@@ -79,10 +88,156 @@ def check_mandatory_settings(mandatory_settings):
+    Gets the branch to use of ocw-hugo-themes in pipelines, defaulting to settings.GITHUB_WEBHOOK_BRANCH
+    if settings.ENVIRONMENT is anything but "dev," otherwise take the value of settings.OCW_HUGO_THEMES_BRANCH
+    """
+    github_webhook_branch = settings.GITHUB_WEBHOOK_BRANCH
```
You mean instead of `settings.GITHUB_WEBHOOK_BRANCH`?  Our Heroku environments currently only have `settings.GITHUB_WEBHOOK_BRANCH` which is set to `release-candidate` in RC and `release` in Production.  In these environments, we want the branch of `ocw-hugo-themes` and `ocw-hugo-projects` to match exactly.  When you're developing locally, you might want to try one branch of `ocw-hugo-themes` with another branch of `ocw-hugo-projects` where you're working on configuration changes to the `config.yaml` file.  This allows you to push up your changes to your development branch, run your pipeline and see the output as built by Concourse.  Most of these scenarios can just be tested by running the hugo / node dev servers in `ocw-hugo-themes`, but with this alternate theme rendering stuff I'm working on I need to be able to test loading a different theme with a different config file and I want to be able to test having Concourse output the site to a different path.

gumaerc commented on gdrive_sync/api.py:
```diff
@@ -201,7 +202,10 @@ def process_file_result(
     """ Stream a Google Drive file to S3 """
     try:
-        s3 = boto3.resource("s3")
+        if settings.ENVIRONMENT == "dev":
```
This was test related.  I tried using simply `get_boto3_resource` here and the tests would fail.  Here's an example of one of the errors:

```
    @pytest.mark.django_db
    @pytest.mark.parametrize("num_errors", [2, 3, 4])
    def test_stream_to_s3_error(settings, mocker, num_errors):
        """Task should mark DriveFile status as failed if an s3 upload error occurs more often than retries"""
        settings.ENVIRONMENT = "test"
        settings.CONTENT_SYNC_RETRIES = 3
        should_raise = num_errors >= 3
        mocker.patch("gdrive_sync.api.GDriveStreamReader")
        mock_boto3 = mocker.patch("gdrive_sync.api.boto3")
        mock_boto3.resource.return_value.Bucket.side_effect = [
            *[HTTPError() for _ in range(num_errors)],
            mocker.Mock(),
        ]
        drive_file = DriveFileFactory.create()
        if should_raise:
            with pytest.raises(HTTPError):
                api.stream_to_s3(drive_file)
        else:
>           api.stream_to_s3(drive_file)

gdrive_sync/api_test.py:189: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
content_sync/decorators.py:29: in wrapper
    return func(*args, **kwargs)
gdrive_sync/api.py:220: in stream_to_s3
    bucket.upload_fileobj(
/usr/local/lib/python3.8/site-packages/boto3/s3/inject.py:580: in bucket_upload_fileobj
    return self.meta.client.upload_fileobj(
/usr/local/lib/python3.8/site-packages/boto3/s3/inject.py:540: in upload_fileobj
    return future.result()
/usr/local/lib/python3.8/site-packages/s3transfer/futures.py:106: in result
    return self._coordinator.result()
/usr/local/lib/python3.8/site-packages/s3transfer/futures.py:265: in result
    raise self._exception
/usr/local/lib/python3.8/site-packages/s3transfer/tasks.py:255: in _main
    self._submit(transfer_future=transfer_future, **kwargs)
/usr/local/lib/python3.8/site-packages/s3transfer/upload.py:554: in _submit
    if not upload_input_manager.requires_multipart_upload(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <s3transfer.upload.UploadSeekableInputManager object at 0x7f2bdd06a310>
transfer_future = <s3transfer.futures.TransferFuture object at 0x7f2bdd06fcd0>, config = <boto3.s3.transfer.TransferConfig object at 0x7f2bdd50a550>

    def requires_multipart_upload(self, transfer_future, config):
>       return transfer_future.meta.size >= config.multipart_threshold
E       TypeError: '>=' not supported between instances of 'MagicMock' and 'int'

/usr/local/lib/python3.8/site-packages/s3transfer/upload.py:240: TypeError
```

I think this is another example of Mocker not behaving.  In in `test_stream_to_s3` we are using `mock_boto3 = mocker.patch("gdrive_sync.api.boto3")` and I don't think that our call to initialize boto3 is properly being mocked because we aren't calling boto3 directly in the scope of the test function.  It seems if the function is called indirectly, `mocker.patch` doesn't work.  I even tried changing it to `mock_boto3 = mocker.patch("main.s3_utils.boto3")` which would hypothetically patch the `boto3` instance being pulled in by the utility function being called by the function the test is calling, and I still got that `TypeError: '>=' not supported between instances of 'MagicMock' and 'int'`.  Any ideas?

gumaerc replied:
Oh, nevermind on this one.  I didn't restart web container after changing the mock target :facepalm:

mbertrand commented on content_sync/pipelines/definitions/concourse/site-pipeline.yml:
```diff
@@ -219,8 +246,8 @@ jobs:
-              aws s3 sync course-markdown/public s3://((ocw-bucket))/((base-url)) --metadata site-id=((site-name))
+              aws s3((cli-endpoint-url)) sync s3://((ocw-studio-bucket))/((s3-path)) s3://((ocw-bucket))/((site-url)) --metadata site-id=((site-name))              
+              aws s3((cli-endpoint-url)) sync course-markdown/public s3://((ocw-bucket))/((base-url)) --metadata site-id=((site-name))
         on_failure:
```
One minor issue I noticed is that there is no webhook sent back on dev when the pipeline completes successfully because that is only included in the `clear-cdn-cache` step.  Could get around that by adding another `DEV-ONLY` section with that webhook call added here just before `on_failure.  

```
        on_success:
          try:
            do:
              - put: ocw-studio-webhook
                timeout: 1m
                attempts: 3
                params:
                  text: |
                    {
                      "version": "((pipeline_name))",
                      "status": "succeeded"
                    }
```                    

<img width="170" alt="Screen Shot 2022-07-27 at 8 45 48 AM" src="https://user-images.githubusercontent.com/187676/181250030-4a8f5ad4-94ad-46aa-9558-cf2e5d1cfed3.png">

mbertrand reviewed (commented):
PS There's also a conflict now in `docker-compose.yml`

gumaerc commented:
@mbertrand Thanks, should be ready for another look.

mbertrand reviewed (approved):
Works great!
